src/main/kotlin/GameOfLife.kt

```diff
@@ -0,0 +1,113 @@
+@file:Suppress("CanSealedSubClassBeObject")
+
+import GenerationCount.Finite
+import GenerationCount.Infinite
+import arrow.core.*
+import arrow.core.extensions.list.functorFilter.filterMap
+import arrow.core.extensions.list.semigroupal.times
+import arrow.fx.ForIO
+import arrow.fx.IO
+import arrow.fx.extensions.io.monad.monad
+import arrow.mtl.StateT
+
+sealed class Cell {
+    class Alive : Cell()
+    class Dead : Cell()
+}
+
+fun Cell.isDead() = !isAlive()
+fun Cell.isAlive() = this is Cell.Alive
+
+typealias Universe = List<List<Cell>>
+
+private val Universe.sizeX: Int get() = size
+
+private val Universe.sizeY: Int get() = get(0).size
+
+data class Position(val x: Int, val y: Int)
+
+fun Option<Position>.shiftBy(universe: Universe, delta: Tuple2<Int, Int>): Option<Position> =
+    flatMap { pos ->
+        val shifted = pos.copy(x = pos.x + delta.a, y = pos.y + delta.b)
+        if (shifted.x in 0 until universe.sizeX && shifted.y in 0 until universe.sizeY) {
+            shifted.some()
+        } else {
+            None
+        }
+    }
+
+fun Universe.cellPosition(cell: Cell): Option<Position> =
+    this.flatten().indexOf(cell).let { index ->
+        if (index == -1) {
+            None
+        } else {
+            val x = index / sizeX
+            val y = index % sizeX
+            Some(Position(x, y))
+        }
+    }
+
+fun Cell.neighbours(universe: Universe): List<Cell> {
+    val deltas = listOf(-1, 0, 1) * listOf(-1, 0, 1) // ListK cartesian product to get all combinations
+    return deltas
+        .filter { it.a != 0 || it.b != 0 }
```

Comment:
I had to think about this part for a few seconds and then I realized that (0,0) is being removed. Maybe there is a way to remove it. I don't know...

Reply:
I rewrote it in 2ab8880 to what I think should be more clear, let me know what you think.

Reply:
I love it!!!!!!!!!!!!